Properly handle list and nil default in ivy-completing-read

DarwinAwardWinner opened the pull request:
This pull request allows `ivy-completing-read` to handle a DEF argument that is a list or nil consistently with `completing-read-default`. It also adds tests for this behavior, including tests for the case where DEF is not an element of COLLECTION.

DarwinAwardWinner commented:
Should fix #1041 and also address #636 even more.

DarwinAwardWinner commented:
By the way, I have prevoiusly completed the Emacs copyright assignment.

abo-abo commented:
Thanks for the work, a large portion of it is definitely an improvement.

But this I don't like:

    (should
     (equal ""
            (ivy-with '(ivy-completing-read "Pick: " '("a" "b" "c") nil t nil nil nil)
                      "RET")))

It means that even this common code will have `""` as the unnecessary and (always) unwanted default:

    (ivy-completing-read "Pick: " '("a" "b" "c"))

Probably the best way to go is to set `ivy-completing-read-default-is-empty-string` to `nil` by default and update the last two tests accordingly.

DarwinAwardWinner commented:
The problem it that it's *not* always unwanted. Some functions call `completing-read` and handle the default themselves if it returns `""`. The docstring for `completing-read` specifically mentions this behavior (emphasis mine):

> If the input is null, ‘completing-read’ returns DEF, or the first element of the list of default values, *or an empty string if DEF is nil, regardless of the value of REQUIRE-MATCH*.

So if you don't allow it to work like this, you'll break some functions that work like this (same example from [here](https://github.com/abo-abo/swiper/issues/636#issuecomment-305069751)) :

```elisp
(defun pick-a-fruit ()
  (interactive)
  (let* ((default "banana")
         (prompt (format "Pick a fruit (default %s): " default))
         (collection '("apple" "banana" "cherry"))
         (selection (completing-read prompt collection nil t)))
    (when (string= selection "")
      (setq selection default))
    (message "You selected %s" selection)))
```

Using ivy would effectively change the default to "apple" rather than "banana", despite the fact that the prompt clearly specifies that the default is banana. You'd have to use `C-M-j` to get "banana". I think it's unfortunate that `completing-read` works in this way, but it can't really be changed now. As a compromise, this behavior could be enabled only when REQUIRE-MATCH is non-nil, since otherwise you can already enter an empty string with `C-M-j`.

Alternatively, you could default `ivy-completing-read-default-is-empty-string` to nil and then implement a whitelist of commands that *do* expect `""` as the implicit default and only set `ivy-completing-read-default-is-empty-string` to non-nil for those commands. You can steal the current list from ido-ubiquitous 3.x. Just pick out all the commands that are marked as `enable-old` from [here](https://github.com/DarwinAwardWinner/ido-ubiquitous/blob/3.x/ido-ubiquitous.el#L306-L371). (This is going away in ido-ubiquitous version 4.x in favor of the behavior described above.)

DarwinAwardWinner commented:
Of course, the other option is to just disable ivy for those commands.

abo-abo commented:
It's possible to do, see `ivy-completing-read-handlers-alist`.

DarwinAwardWinner commented:
Ok, instead of having a variable to control default handling, I just implemented a new function called `ivy-completing-read-with-empty-string-def` which implements the compatible behavior, while leaving `ivy-completing-read` alone. Then I added some entries to `ivy-completing-read-handlers-alist` using this function, whereupon I realized that having a handler that itself calls `ivy-completing-read` causes infinite recursion. So I fixed that as well. Lastly, I added a test for `ivy-completing-read-handlers-alist`.

With these changes, ivy should now work with all of the functions that I know of that expect this special behavior from `completing-read` (the offending packages are listed in my comment [here](https://github.com/abo-abo/swiper/issues/636#issuecomment-305069751)).

DarwinAwardWinner commented:
According to the CI tests, Emacs 24.3 doesn't seem to like `cl-letf*`. I'll investigate this.

DarwinAwardWinner commented:
The root cause was a bug in `symbol-function` that was fixed in 24.4. I backported the fix just for the test, and now everything passes.

abo-abo commented:
Thanks for your work. Very detailed and thorough:)

basil-conto commented on ivy.el:
```diff
@@ -1879,6 +1897,21 @@ INHERIT-INPUT-METHOD is currently ignored."
+Specifically, if DEF is nil, it is treated the same as if DEF was
+the empty string. This mimics the behavior of
+`completing-read-refault'. This function can therefore be used in
+place of `icy-completing-read' for commands that rely on this
```
`s/icy/ivy/`

abo-abo replied:
haha, I'll fix it later
